# **Finding and Removing Duplicate Data w/ Pandas**

**In this tutorial, I'll show an example of identifying duplicated rows within a dataframe and how to remove those rows using Pandas.**

In [1]:
import pandas as pd

**We'll use data border crossing data released by the Bureau of Transportation Statistics. The dataset can be found on Kaggle.**

In [2]:
file = "../Border_Crossing_Project/Border_Crossing_Entry_Data.csv"

In [3]:
df = pd.read_csv(file, parse_dates=['Date'])
df.head()

Port Name State  Port Code            Border       Date  \
0     Alcan    AK       3104  US-Canada Border 2020-02-01   
1     Alcan    AK       3104  US-Canada Border 2020-02-01   
2     Alcan    AK       3104  US-Canada Border 2020-02-01   
3     Alcan    AK       3104  US-Canada Border 2020-02-01   
4     Alcan    AK       3104  US-Canada Border 2020-02-01   

                       Measure  Value  
0  Personal Vehicle Passengers   1414  
1            Personal Vehicles    763  
2       Truck Containers Empty    412  
3        Truck Containers Full    122  
4                       Trucks    545

In [4]:
df.shape

(355511, 7)

**After converting the csv to a pandas dataframe, we can use the "duplicated" method to return a boolean series that identifies any duplicated rows as True, or otherwise False.**

In [5]:
df.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
355506    False
355507    False
355508    False
355509    False
355510    False
Length: 355511, dtype: bool

**Instead of just returning a series, we can add this as a column to the dataframe.**

In [6]:
df['duplicated'] = df.duplicated()
df.head()

Port Name State  Port Code            Border       Date  \
0     Alcan    AK       3104  US-Canada Border 2020-02-01   
1     Alcan    AK       3104  US-Canada Border 2020-02-01   
2     Alcan    AK       3104  US-Canada Border 2020-02-01   
3     Alcan    AK       3104  US-Canada Border 2020-02-01   
4     Alcan    AK       3104  US-Canada Border 2020-02-01   

                       Measure  Value  duplicated  
0  Personal Vehicle Passengers   1414       False  
1            Personal Vehicles    763       False  
2       Truck Containers Empty    412       False  
3        Truck Containers Full    122       False  
4                       Trucks    545       False

**Here we return just the rows that were identified as duplicates.**

In [7]:
df.loc[df['duplicated'] == True]

Port Name State  Port Code            Border       Date  \
6590       Del Rio    TX       2302  US-Mexico Border 2019-06-01   
6591       Del Rio    TX       2302  US-Mexico Border 2019-06-01   
6977       Detroit    MI       3801  US-Canada Border 2019-06-01   
7129  Santa Teresa    NM       2408  US-Mexico Border 2019-06-01   
7130  Santa Teresa    NM       2408  US-Mexico Border 2019-06-01   
7133          Roma    TX       2310  US-Mexico Border 2019-06-01   
7134          Roma    TX       2310  US-Mexico Border 2019-06-01   
7137     Fort Kent    ME        110  US-Canada Border 2019-06-01   
7138     Fort Kent    ME        110  US-Canada Border 2019-06-01   
7141       Detroit    MI       3801  US-Canada Border 2019-06-01   

                          Measure   Value  duplicated  
6590            Personal Vehicles  123068        True  
6591  Personal Vehicle Passengers  238721        True  
6977  Personal Vehicle Passengers  599521        True  
7129            Personal Vehicles   50672        True  
7130  Personal Vehicle Passengers  110576        True  
7133            Personal Vehicles   54953        True  
7134  Personal Vehicle Passengers  104640        True  
7137            Personal Vehicles    9615        True  
7138  Personal Vehicle Passengers   14056        True  
7141            Personal Vehicles  351889        True

**Let's look at an example of one of the duplicates to confirm that we are indeed identifying duplicate rows, but allowing us to keep one by labeling its duplicated value as False. Since we know the rows must be identical, let's search based on the 'Value' column since it is most likely a value in that column will be unique to the dataset. Here we'll search for the row(s) with a value of '123068', the first row in the duplicates table above.**

In [8]:
df.loc[df['Value'] == 123068]

Port Name State  Port Code            Border       Date  \
6362   Del Rio    TX       2302  US-Mexico Border 2019-06-01   
6590   Del Rio    TX       2302  US-Mexico Border 2019-06-01   

                Measure   Value  duplicated  
6362  Personal Vehicles  123068       False  
6590  Personal Vehicles  123068        True

**As expected, there are duplicated rows w/ this value, in this case just one extra row. It's duplicated value is set to False so it can be retained during duplicate removal.**

**The duplicated method has a parameter, 'keep', which can be used to modify which duplicated values are marked as True. The default is 'first', meaning all but the first occurence are marked as duplicates. Other options are 'last', to mark all but the last occurence as duplicates, and False, which will mark all occurences as duplicates.**




**Now that we have identified that there are duplicates in our dataset, we can proceed w/ removing this data.**
**Let's print the number of the rows of the original dataset along w/ the number of duplicated rows.**

In [9]:
print(f"The number of rows in the dataframe is: {len(df)}")
print(f"The number of duplicated rows is: {len(df.loc[df['duplicated'] == True])}")

The number of rows in the dataframe is: 355511
The number of duplicated rows is: 10


**We see that there are 355,511 rows in the dataframe and 10 of those are duplicates so we expect the number of rows to be 355,501 after removing the duplicates.**

**Pandas provides a method, 'drop_duplicates()', that will remove the duplicated rows. Just like the 'duplicated()' method, it has a 'keep' parameter w/ the same options. We'll leave it at the default, which is 'first'.** 

In [10]:
df_dupes_dropped = df.drop_duplicates()
df_dupes_dropped.head()

Port Name State  Port Code            Border       Date  \
0     Alcan    AK       3104  US-Canada Border 2020-02-01   
1     Alcan    AK       3104  US-Canada Border 2020-02-01   
2     Alcan    AK       3104  US-Canada Border 2020-02-01   
3     Alcan    AK       3104  US-Canada Border 2020-02-01   
4     Alcan    AK       3104  US-Canada Border 2020-02-01   

                       Measure  Value  duplicated  
0  Personal Vehicle Passengers   1414       False  
1            Personal Vehicles    763       False  
2       Truck Containers Empty    412       False  
3        Truck Containers Full    122       False  
4                       Trucks    545       False

**Let's check the length of the new dataframe to see if we correctly removed the duplicated rows.**

In [11]:
print(f"The number of rows in the dataframe w/ duplicates removed is: {len(df_dupes_dropped)}")

The number of rows in the dataframe w/ duplicates removed is: 355511


      
    
**Interesting, we still have the same number of rows. How could that be?** 
  
**Remember when we added the duplicated column to the dataframe? Well that made no rows duplicates because in this case there is only one duplicate for each case of duplication and thus after adding a column of True/False, we effectively made all rows unique.**  
  
**So let's remove that column and then rerun the process.**

In [12]:
df_dupes_dropped = df.drop(columns=['duplicated']).drop_duplicates()

In [15]:
df_dupes_dropped.head()

Port Name State  Port Code            Border       Date  \
0     Alcan    AK       3104  US-Canada Border 2020-02-01   
1     Alcan    AK       3104  US-Canada Border 2020-02-01   
2     Alcan    AK       3104  US-Canada Border 2020-02-01   
3     Alcan    AK       3104  US-Canada Border 2020-02-01   
4     Alcan    AK       3104  US-Canada Border 2020-02-01   

                       Measure  Value  
0  Personal Vehicle Passengers   1414  
1            Personal Vehicles    763  
2       Truck Containers Empty    412  
3        Truck Containers Full    122  
4                       Trucks    545

In [13]:
print(f"The number of rows in the dataframe w/ duplicates removed is: {len(df_dupes_dropped)}")

The number of rows in the dataframe w/ duplicates removed is: 355501


**Here we see the result we expected. Let's just confirm by returning the unique values of running 'duplicated()' on the new dataframe, which should return only False.**

In [14]:
df_dupes_dropped.duplicated().unique()

array([False])

In [17]:
df_dupes_dropped.to_csv('Border_Crossing_Data_no_duplicates.csv', index=False)